test/assert_checkcircuit.go

```diff
@@ -0,0 +1,285 @@
+package test
+
+import (
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/backend/plonk"
+	"github.com/consensys/gnark/backend/plonkfri"
+	"github.com/consensys/gnark/backend/witness"
+	"github.com/consensys/gnark/constraint"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/schema"
+)
+
+// CheckCircuit performs a series of check on the provided circuit.
+//
+//	go test -short 					--> testEngineChecks
+//	go test 						--> testEngineChecks  + constraintSolverChecks
+//	go test -tags=prover_checks 	--> ... + proverChecks
+//	go test -tags=release_checks 	--> ... + releaseChecks (solidity, serialization, ...)
+//
+// Depending on the above flags, the following checks are performed:
+// - the circuit compiles
+// - the circuit can be solved with the test engine
+// - the circuit can be solved with the constraint system solver
+// - the circuit can be solved with the prover
+// - the circuit can be verified with the verifier
+// - the circuit can be verified with gnark-solidity-checker
+// - the circuit, witness, proving and verifying keys can be serialized and deserialized
+func (assert *Assert) CheckCircuit(circuit frontend.Circuit, opts ...TestingOption) {
+	// get the testing configuration
+	opt := assert.options(opts...)
+
+	// for each {curve, backend} tuple
+	for _, curve := range opt.curves {
+		curve := curve
+
+		// run in sub-test to contextualize with curve
+		assert.Run(func(assert *Assert) {
+
+			// parse valid / invalid assignments
+			var invalidWitnesses, validWitnesses []_witness
+			for _, a := range opt.validAssignments {
+				w := assert.parseAssignment(circuit, a, curve, opt.checkSerialization)
+				validWitnesses = append(validWitnesses, w)
+
+				// check that the assignment is valid with the test engine
+				err := IsSolved(circuit, w.assignment, curve.ScalarField())
```

Comment:
Another use case for avoiding test engine - when I'm trying to debug hint functions (particularly a case where I replace a standard hint with a malicious one), then because test engine does not use hint registry but instead calls the hint functions directly, then I'm not able to override hint function in the test engine.

And I'm not able to make tests pass as I'm not getting to solver as the test will fail with the test engine (but should succeed with solver).